### PR TITLE
Policy for download restriction

### PR DESCRIPTION
**Pull request by joshmoore**
See: https://trello.com/c/Y1YocYjU/323-download-button-restriction

Add support for limiting the download of original files. The basic functionality is that it is now possible to configure the server to limit access to original files, for example, preventing downloading the fileset.

## Client-side: Add Permissions.extendedRestrictions mechanism

In order to allow further canEdit/canDelete/...-like
functionality, a second field on Permissions has been
added which contains strings documented and set by
individual services at runtime. Failure to obey these
restrictions in the client will lead to an exception.

## Server-side: ome.security.policy.PolicyService

In order to support the new Permissions.extendedRestriction
client-side checks, this commit adds a server-side interface,
ome.security.policy.PolicyService, which is configured with
instances of ome.security.policy.Policy. Similar to the JVM's
SecurityManager and Permission objects, calling code can either
check if a call *would* fail, or simply let the service raise
a SecurityViolation.

Currently, the only prototypical restriction policy in place
is the DownloadPolicy.

cc: @mtbc @aleksandra-tarkowska @chris-allan @imunro 

## Testing

 * Check that with this PR merged but no special configuration that all downloads continue to work
 * Disable all downloads:
   * `bin/omero config set omero.policy.download none`
   * check that **nothing** is downloadable
   * e.g. `bin/omero upload some_file.txt && bin/omero download 32 -`
   * note: **import** will also be broken. It may be that this is a useless config.
 * Set the repository setting:
   * `bin/omero config set omero.policy.download repository`
   * check that the owner of a file can download it
   * but others in a `rwr---` or `rwra--` group can **not**

### CODE REVIEW

File: components/server/src/ome/security/policy/PolicyService.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+     * Return all configured identifier strings as would be passed as the first
+     * argument to {@link #isRestricted(String, IObject)} or
+     * {@link #checkRestriction(String, IObject)}.
+     */
+    Set<String> listAllRestrictions();
+
+    /**
+     * Return all identifier strings as would be passed as the first argument to
+     * {@link #isRestricted(String, IObject)} or
+     * {@link #checkRestriction(String, IObject)} <em>which</which> considers
```
Comment (chris-allan) [inline components/server/src/ome/security/policy/PolicyService.java]: `</which>`?

File: components/tools/OmeroPy/src/omero/plugins/download.py [hunk 2]
```diff
@@ -70,14 +72,14 @@ def __call__(self, args):
             # ID exists in DB, but not on FS
             self.ctx.die(67, "ResourceError: %s" % re.message)
 
-    def get_file_id(self, session, value):
+    def get_filed(self, session, value):
```
Comment (chris-allan) [inline components/tools/OmeroPy/src/omero/plugins/download.py]: ?

**Comment by aleksandra-tarkowska** [comment]
Just briefly tested:
 - unset omero.policy.download works fine.
 - bin/omero config set  omero.policy.download true

  ```
<class 'omeroweb.webclient.webclient_gateway.ImageWrapper'> False
<class 'omero.gateway.FileAnnotationWrapper'> True
```
is file annotation meant to be considered?

**Comment by aleksandra-tarkowska** [comment]
recently @will-moore ported PR that allow SPW restriction https://github.com/joshmoore/openmicroscopy/blob/restrict-download/components/tools/OmeroWeb/omeroweb/webclient/controller/container.py#L206 Just discussed with @chris-allan the PLATE_DOWNLOAD_ENABLED is not ideal and download.limit is definitely would be better. 
```
Chris Allan
It also then applies to all of these cases where we have people screwing up their installs by doing things like trying to download a 10GB raw file and a 1KB companion file that results in huge ZIP file downloads, etc. etc.
The same is true for the select 50 images for download and now I have 10+GB ZIP file download. There are lots of situations where you can really cause yourself serious trouble right now and all we're protecting ourselves from right now is SPW download.```

**Comment by joshmoore** [comment]
Pushed @aleksandra-tarkowska's commits as well as a minor change to how the supported types work. At this point, Plate, Well, Image, OriginalFile, and FileAnnotation should all be getting the `DOWNLOAD` restriction, but *only* based on whether or not the current user can edit them.

I'll need to re-work how the transactions are happening in order to base the policy decision on more information (contents of the plate, size of the file, etc.) as well as to make the decision based on the group rather than the server config. That may should happen in a second PR.

**Comment by aleksandra-tarkowska** [comment]
Tested locally. All works as expected. One issue noticed:
in ```repository``` mode ```rwra--``` group only member couldn't download the data. Admin and group owner could. I think group owner should be restricted the same as user. Admins are different stories...

**Comment by pwalczysko** [comment]
Unrelated problem with downloads - discussed with @will-moore : Created https://trac.openmicroscopy.org/ome/ticket/12776.

**Comment by joshmoore** [comment]
> I think group owner should be restricted the same as user.

group owners can sudo to group members, therefore preventing download doesn't really give you any protection.

**Comment by pwalczysko** [comment]
Downloads icon missing in Insight - @dominikl is working on this (unrelated to this PR).

**Comment by dominikl** [comment]
@pwalczysko Just openend a PR #3571 , download button should be back again in tomorrows build. Sorry, should have done that already yesterday, when I noticed the missing download button, too.

File: components/blitz/resources/omero/model/Permissions.ice [hunk 3]
```diff
@@ -37,6 +37,17 @@ module omero {
       omero::api::BoolArray restrictions;
 
       /**
+       * Further restrictions which are specified by services
+       * at runtime. Individual service methods will specify
+       * which strings MAY NOT be present in this field for
+       * execution to be successful. For example, if an
+       * [omero::model::Image] contains a "DOWNLOAD" restriction,
+       * then attempt to call [omero::api::RawFileStore::read] will
```
Comment (mtbc) [inline components/blitz/resources/omero/model/Permissions.ice]: "an attempt" / "attempts"

File: components/server/src/ome/security/policy/Policy.java [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+ *    }
+ * </pre>
+ */
+public interface Policy {
+
+    /**
+     * Unique name for this type of {@link Policy}. This string will be sent to
+     * clients via
+     * {@link ome.model.internal.Permissions#copyExtendedRestrictions()} in
+     * order to prevent exceptions.
```
Comment (mtbc) [inline components/server/src/ome/security/policy/Policy.java]: Not sure I have anything to suggest here, but this made sense only when I realized that it is the policy names that get passed as the extended restrictions, it's more than a name to put in a log message or whatever.

**Comment by pwalczysko** [comment]
when the buttons are back, it would be good to test against local.

File: components/server/src/ome/security/policy/DefaultPolicyService.java [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+        }
+    }
+
+    //
+    // INTERFACE METHODS
+    //
+
+    @Override
+    public boolean isRestricted(final String name, final IObject obj) {
+        for (Policy check : policies.get(name))
```
Comment (mtbc) [inline components/server/src/ome/security/policy/DefaultPolicyService.java]: Could be null?

**Comment by mtbc** [comment]
I am guessing that this is still work in progress but, at least focusing on the API, this is already looking pretty good.

**Comment by aleksandra-tarkowska** [comment]
@pwalczysko noticed that download button on well is available, I will resolve that

**Comment by sbesson** [comment]
Is commit  f0fcba2 causing the failure of https://ci.openmicroscopy.org/view/Failing/job/OMERO-5.1-merge-integration-web/lastCompletedBuild/testReport/test.integration.test_download/TestDownload/test_spw_download/. This test was also modified as part of https://github.com/openmicroscopy/openmicroscopy/pull/3564 but I cannot reproduce locally from my branch and this commit seems to have implication on the return code.

**Comment by aleksandra-tarkowska** [comment]
Yes, working on it

**Comment by pwalczysko** [comment]
@joshmoore @aleksandra-tarkowska @will-moore @dominikl : I will just try to recapitulate:

   - 1. we have a problem with Insight download button (client side only, but not helping the overview here)
   - 2. download button on plates is available in web, although it should not (as a default) 
   - 3. the actual download of plates works as a default, although it should not 
   - I think I will retest tomorrow. Hopefully this will be clearer here.

**Comment by aleksandra-tarkowska** [comment]
@joshmoore when bin/omero config set ```omero.policy.download none```:
```
omero.SecurityViolation
    serverStackTrace = "ome.conditions.SecurityViolation: Download is restricted for ome.model.core.OriginalFile:Id_1773
                        	at ome.security.policy.DownloadPolicy.checkRestriction(DownloadPolicy.java:88)
                        	at ome.security.policy.DefaultPolicyService.checkRestriction(DefaultPolicyService.java:77)
                        	at ome.security.basic.BasicSecuritySystem.checkRestriction(BasicSecuritySystem.java:652)
                        	at ome.services.RawFileBean.setFileIdWithBuffer(RawFileBean.java:346)
                        	at ome.services.blitz.repo.RepositoryDaoImpl$1.doWork(RepositoryDaoImpl.java:172)
                        	at sun.reflect.GeneratedMethodAccessor274.invoke(Unknown Source)
                        	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
                        	at java.lang.reflect.Method.invoke(Method.java:483)
                        	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:307)
                        	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:183)
                        	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:150)
                        	at ome.services.util.Executor$Impl$Interceptor.invoke(Executor.java:566)
                        	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
                        	at ome.security.basic.EventHandler.invoke(EventHandler.java:154)
                        	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
                        	at ome.tools.hibernate.SessionHandler.doStateful(SessionHandler.java:218)
                        	at ome.tools.hibernate.SessionHandler.invoke(SessionHandler.java:202)
                        	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
                        	at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:108)
                        	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
                        	at ome.tools.hibernate.ProxyCleanupFilter$Interceptor.invoke(ProxyCleanupFilter.java:241)
                        	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
                        	at ome.services.util.ServiceHandler.invoke(ServiceHandler.java:116)
                        	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
                        	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:202)
                        	at com.sun.proxy.$Proxy74.doWork(Unknown Source)
                        	at ome.services.util.Executor$Impl.execute(Executor.java:447)
                        	at ome.services.blitz.repo.RepositoryDaoImpl.getRawFileStore(RepositoryDaoImpl.java:167)
                        	at ome.services.blitz.repo.PublicRepositoryI.createRepoRFS(PublicRepositoryI.java:601)
                        	at ome.services.blitz.repo.PublicRepositoryI.file(PublicRepositoryI.java:413)
                        	at ome.services.blitz.repo.ManagedImportProcessI.getUploader(ManagedImportProcessI.java:224)
                        	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
                        	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
                        	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
                        	at java.lang.reflect.Method.invoke(Method.java:483)
                        	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:307)
                        	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:183)
                        	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:150)
                        	at omero.cmd.CallContext.invoke(CallContext.java:78)
                        	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
                        	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:202)
                        	at com.sun.proxy.$Proxy106.getUploader(Unknown Source)
                        	at omero.grid._ImportProcessTie.getUploader(_ImportProcessTie.java:107)
                        	at omero.grid._ImportProcessDisp.___getUploader(_ImportProcessDisp.java:204)
                        	at omero.grid._ImportProcessDisp.__dispatch(_ImportProcessDisp.java:345)
                        	at IceInternal.Incoming.invoke(Incoming.java:222)
                        	at Ice.ConnectionI.invokeAll(ConnectionI.java:2482)
                        	at Ice.ConnectionI.dispatch(ConnectionI.java:1258)
                        	at Ice.ConnectionI.message(ConnectionI.java:1213)
                        	at IceInternal.ThreadPool.run(ThreadPool.java:321)
                        	at IceInternal.ThreadPool.access$300(ThreadPool.java:12)
                        	at IceInternal.ThreadPool$EventHandlerThread.run(ThreadPool.java:693)
                        	at java.lang.Thread.run(Thread.java:745)
                        "
    serverExceptionClass = "ome.conditions.SecurityViolation"
    message = "Download is restricted for ome.model.core.OriginalFile:Id_1773"
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:408)
	at java.lang.Class.newInstance(Class.java:438)
	at IceInternal.BasicStream.createUserException(BasicStream.java:2615)
	at IceInternal.BasicStream.access$300(BasicStream.java:12)
	at IceInternal.BasicStream$EncapsDecoder10.throwException(BasicStream.java:3099)
	at IceInternal.BasicStream.throwException(BasicStream.java:2077)
	at IceInternal.Outgoing.throwUserException(Outgoing.java:538)
	at omero.grid._ImportProcessDelM.getUploader(_ImportProcessDelM.java:352)
	at omero.grid.ImportProcessPrxHelper.getUploader(ImportProcessPrxHelper.java:1680)
	at omero.grid.ImportProcessPrxHelper.getUploader(ImportProcessPrxHelper.java:1642)
	at ome.formats.importer.transfers.TransferState.getUploader(TransferState.java:208)
	at ome.formats.importer.transfers.TransferState.getUploader(TransferState.java:195)
	at ome.formats.importer.transfers.AbstractFileTransfer.start(AbstractFileTransfer.java:100)
	at ome.formats.importer.transfers.UploadFileTransfer.transfer(UploadFileTransfer.java:46)
	at ome.formats.importer.ImportLibrary.uploadFile(ImportLibrary.java:430)
	at org.openmicroscopy.shoola.env.data.OMEROGateway.importImageFile(OMEROGateway.java:6685)
	at org.openmicroscopy.shoola.env.data.OmeroImageServiceImpl.importCandidates(OmeroImageServiceImpl.java:231)
	at org.openmicroscopy.shoola.env.data.OmeroImageServiceImpl.importFile(OmeroImageServiceImpl.java:1274)
	at org.openmicroscopy.shoola.env.data.views.calls.ImagesImporter.importFile(ImagesImporter.java:80)
	at org.openmicroscopy.shoola.env.data.views.calls.ImagesImporter.access$000(ImagesImporter.java:55)
	at org.openmicroscopy.shoola.env.data.views.calls.ImagesImporter$1.doCall(ImagesImporter.java:104)
	at org.openmicroscopy.shoola.env.data.views.BatchCall.doStep(BatchCall.java:144)
	at org.openmicroscopy.shoola.util.concur.tasks.CompositeTask.doStep(CompositeTask.java:226)
	at org.openmicroscopy.shoola.env.data.views.CompositeBatchCall.doStep(CompositeBatchCall.java:126)
	at org.openmicroscopy.shoola.util.concur.tasks.ExecCommand.exec(ExecCommand.java:165)
	at org.openmicroscopy.shoola.util.concur.tasks.ExecCommand.run(ExecCommand.java:276)
	at org.openmicroscopy.shoola.util.concur.tasks.AsyncProcessor$Runner.run(AsyncProcessor.java:91)
	at java.lang.Thread.run(Thread.java:745)

	at org.openmicroscopy.shoola.env.data.OMEROGateway.importImageFile(OMEROGateway.java:6716)
	at org.openmicroscopy.shoola.env.data.OmeroImageServiceImpl.importCandidates(OmeroImageServiceImpl.java:231)
	at org.openmicroscopy.shoola.env.data.OmeroImageServiceImpl.importFile(OmeroImageServiceImpl.java:1274)
	at org.openmicroscopy.shoola.env.data.views.calls.ImagesImporter.importFile(ImagesImporter.java:80)
	at org.openmicroscopy.shoola.env.data.views.calls.ImagesImporter.access$000(ImagesImporter.java:55)
	at org.openmicroscopy.shoola.env.data.views.calls.ImagesImporter$1.doCall(ImagesImporter.java:104)
	at org.openmicroscopy.shoola.env.data.views.BatchCall.doStep(BatchCall.java:144)
	at org.openmicroscopy.shoola.util.concur.tasks.CompositeTask.doStep(CompositeTask.java:226)
	at org.openmicroscopy.shoola.env.data.views.CompositeBatchCall.doStep(CompositeBatchCall.java:126)
	at org.openmicroscopy.shoola.util.concur.tasks.ExecCommand.exec(ExecCommand.java:165)
	at org.openmicroscopy.shoola.util.concur.tasks.ExecCommand.run(ExecCommand.java:276)
	at org.openmicroscopy.shoola.util.concur.tasks.AsyncProcessor$Runner.run(AsyncProcessor.java:91)
	at java.lang.Thread.run(Thread.java:745)
Caused by: omero.SecurityViolation
    serverStackTrace = "ome.conditions.SecurityViolation: Download is restricted for ome.model.core.OriginalFile:Id_1773
                        	at ome.security.policy.DownloadPolicy.checkRestriction(DownloadPolicy.java:88)
                        	at ome.security.policy.DefaultPolicyService.checkRestriction(DefaultPolicyService.java:77)
                        	at ome.security.basic.BasicSecuritySystem.checkRestriction(BasicSecuritySystem.java:652)
                        	at ome.services.RawFileBean.setFileIdWithBuffer(RawFileBean.java:346)
                        	at ome.services.blitz.repo.RepositoryDaoImpl$1.doWork(RepositoryDaoImpl.java:172)
                        	at sun.reflect.GeneratedMethodAccessor274.invoke(Unknown Source)
                        	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
                        	at java.lang.reflect.Method.invoke(Method.java:483)
                        	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:307)
                        	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:183)
                        	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:150)
                        	at ome.services.util.Executor$Impl$Interceptor.invoke(Executor.java:566)
                        	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
                        	at ome.security.basic.EventHandler.invoke(EventHandler.java:154)
                        	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
                        	at ome.tools.hibernate.SessionHandler.doStateful(SessionHandler.java:218)
                        	at ome.tools.hibernate.SessionHandler.invoke(SessionHandler.java:202)
                        	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
                        	at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:108)
                        	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
                        	at ome.tools.hibernate.ProxyCleanupFilter$Interceptor.invoke(ProxyCleanupFilter.java:241)
                        	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
                        	at ome.services.util.ServiceHandler.invoke(ServiceHandler.java:116)
                        	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
                        	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:202)
                        	at com.sun.proxy.$Proxy74.doWork(Unknown Source)
                        	at ome.services.util.Executor$Impl.execute(Executor.java:447)
                        	at ome.services.blitz.repo.RepositoryDaoImpl.getRawFileStore(RepositoryDaoImpl.java:167)
                        	at ome.services.blitz.repo.PublicRepositoryI.createRepoRFS(PublicRepositoryI.java:601)
                        	at ome.services.blitz.repo.PublicRepositoryI.file(PublicRepositoryI.java:413)
                        	at ome.services.blitz.repo.ManagedImportProcessI.getUploader(ManagedImportProcessI.java:224)
                        	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
                        	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
                        	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
                        	at java.lang.reflect.Method.invoke(Method.java:483)
                        	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:307)
                        	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:183)
                        	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:150)
                        	at omero.cmd.CallContext.invoke(CallContext.java:78)
                        	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
                        	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:202)
                        	at com.sun.proxy.$Proxy106.getUploader(Unknown Source)
                        	at omero.grid._ImportProcessTie.getUploader(_ImportProcessTie.java:107)
                        	at omero.grid._ImportProcessDisp.___getUploader(_ImportProcessDisp.java:204)
                        	at omero.grid._ImportProcessDisp.__dispatch(_ImportProcessDisp.java:345)
                        	at IceInternal.Incoming.invoke(Incoming.java:222)
                        	at Ice.ConnectionI.invokeAll(ConnectionI.java:2482)
                        	at Ice.ConnectionI.dispatch(ConnectionI.java:1258)
                        	at Ice.ConnectionI.message(ConnectionI.java:1213)
                        	at IceInternal.ThreadPool.run(ThreadPool.java:321)
                        	at IceInternal.ThreadPool.access$300(ThreadPool.java:12)
                        	at IceInternal.ThreadPool$EventHandlerThread.run(ThreadPool.java:693)
                        	at java.lang.Thread.run(Thread.java:745)
                        "
    serverExceptionClass = "ome.conditions.SecurityViolation"
    message = "Download is restricted for ome.model.core.OriginalFile:Id_1773"
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:408)
	at java.lang.Class.newInstance(Class.java:438)
	at IceInternal.BasicStream.createUserException(BasicStream.java:2615)
	at IceInternal.BasicStream.access$300(BasicStream.java:12)
	at IceInternal.BasicStream$EncapsDecoder10.throwException(BasicStream.java:3099)
	at IceInternal.BasicStream.throwException(BasicStream.java:2077)
	at IceInternal.Outgoing.throwUserException(Outgoing.java:538)
	at omero.grid._ImportProcessDelM.getUploader(_ImportProcessDelM.java:352)
	at omero.grid.ImportProcessPrxHelper.getUploader(ImportProcessPrxHelper.java:1680)
	at omero.grid.ImportProcessPrxHelper.getUploader(ImportProcessPrxHelper.java:1642)
	at ome.formats.importer.transfers.TransferState.getUploader(TransferState.java:208)
	at ome.formats.importer.transfers.TransferState.getUploader(TransferState.java:195)
	at ome.formats.importer.transfers.AbstractFileTransfer.start(AbstractFileTransfer.java:100)
	at ome.formats.importer.transfers.UploadFileTransfer.transfer(UploadFileTransfer.java:46)
	at ome.formats.importer.ImportLibrary.uploadFile(ImportLibrary.java:430)
	at org.openmicroscopy.shoola.env.data.OMEROGateway.importImageFile(OMEROGateway.java:6685)
	... 12 more
```

**Comment by joshmoore** [comment]
* Moved the DOWNLOAD check in RawFileBean to the read rather than setFileId to fix import on `none`. It might be that `none` is just not useful and that we want to either have the default or `repository`, in which case this could move back.
 * Fixed the problem of Plates et al. not having `isRestricted(DOWNLOAD)` return `True` (i.e. in the UI they should not appear). @aleksandra-tarkowska : this might could make your hacks unnecessary, I'm unsure. 

Also : any potential of getting UI tests? My integration tests I've made conditional based on the server configuration. `if not repository: skip()`. Note sure if something like that is possible in the robot tests. cc: @sbesson

**Comment by aleksandra-tarkowska** [comment]
I am not sure if I understand that correctly. If ```bin/omero config set  omero.policy.download```  ```well.getDetails().getPermissions().isRestricted(omero.constants.permissions.DOWNLOAD)``` return ```False```

**Comment by aleksandra-tarkowska** [comment]
all is working, good to merge

final UI fixes and restrictions to SPW specific policies will be done in a separate PR

**Comment by aleksandra-tarkowska** [comment]
@joshmoore just a small comment in terms of setting ```omero.policy.download_size 100MB``` I understand that value will be tested for FilesetEntry size on both Image and Well/Plate objects right?

**Comment by joshmoore** [comment]
> all is working, good to merge

Thanks, Ola.

> I understand that value will be tested for FilesetEntry size on both Image and Well/Plate objects right?

I was wondering that, too. Best to discuss and work through on the next PR.

Summarizing this PR:
 * this PR primarily adds the ability to add a restriction flag
 * the only one currently implemented is the generic "DownloadPolicy"
 * this does *not* yet prevent downloads based on size and/or group. Those and other suggestions will come next.

If there are no further comments, good to merge. @pwalczysko , @jburel , etc. ?

**Comment by pwalczysko** [comment]
No objections. We will go through the download workflows next week.

**Comment by jburel** [comment]
One point to consider later the download of scripts.

**Comment by jburel** [comment]
Merging.

**Pull request merged**